src/ReinforcementLearningZoo/src/algorithms/dqns/NFQ.jl

```diff
@@ -0,0 +1,80 @@
+export NFQ
+
+using Flux
+using Functors: @functor
+
+"""
+    NFQ{A<:AbstractApproximator, F, R} <: AbstractLearner
+    NFQ(approximator::A, num_iterations::Integer epochs::Integer, loss_function::F, batch_size::Integer, rng::R, γ::Float32) where {A<:AbstractApproximator, F, R}
+Neural Fitted Q-iteration as implemented in [1]
+
+# Keyword arguments
+- `approximator::AbstractApproximator` neural network
+- `num_iterations::Integer` number of value iteration iterations in FQI loop (i.e. the outer loop)
+- `epochs` number of epochs to train neural network per iteration
+- `loss_function::F` loss function of the NN
+- `sampler::BatchSampler{SARTS}` data sampler
+- `rng::R` random number generator
+- `γ::Float32` discount rate
+
+# References
+[1] Riedmiller, M. (2005). Neural Fitted Q Iteration – First Experiences with a Data Efficient Neural Reinforcement Learning Method. In: Gama, J., Camacho, R., Brazdil, P.B., Jorge, A.M., Torgo, L. (eds) Machine Learning: ECML 2005. ECML 2005. Lecture Notes in Computer Science(), vol 3720. Springer, Berlin, Heidelberg. https://doi.org/10.1007/11564096_32
+"""
+Base.@kwdef struct NFQ{A, R, F} <: AbstractLearner
+    action_space::AbstractVector
+    approximator::A
+    num_iterations::Integer = 20
+    epochs::Integer = 100
+    loss_function::F = mse
+    rng::R = Random.default_rng()
+    γ::Float32 = 0.9f0
+end
+
+# function NFQ(;
+#     action_space::AbstractVector,
+#     approximator::A,
+#     num_iterations::Integer = 20,
+#     epochs::Integer = 1000,
+#     loss_function::Any = mse,
+#     rng=Random.default_rng(),
+#     γ::Float32 = 0.9f0,
+#     ) where {A}
+#     NFQ(action_space, approximator, num_iterations, epochs, loss_function, rng, γ)
+# end
+
+# Copied from BasicDQN but sure whether it's appropriate
+@functor NFQ (approximator,)
+
+RLCore.forward(L::NFQ, s::AbstractArray) = RLCore.forward(L.approximator, s)
+
+function RLCore.forward(learner::NFQ, env::AbstractEnv)
+    as = action_space(env)
+    return vcat(repeat(state(env), inner=(1, length(as))), transpose(as)) |> x -> send_to_device(device(learner.approximator), x) |> x->RLCore.forward(learner, x) |> send_to_host |> vec 
+end
+
+# Avoid optimisation in the middle of an episode
+function RLBase.optimise!(::NFQ, ::AbstractStage, ::Trajectory) end
+
+function RLBase.optimise!(learner::NFQ, ::PostEpisodeStage, trajectory::Trajectory)
+    Q = learner.approximator
+    γ = learner.γ
+    loss_func = learner.loss_function
+    as = learner.action_space
+    las = length(as)
+    batch = nothing
+    for b in trajectory
+        batch = b
+    end
+    batch === nothing && return
```

Comment:
I see why you are doing this but it is not how a Trajectory is supposed to be used. Make a `for batch in trajectory` loop around the code below and it should be fine. Because here you are sampling potentially 200 times (for a cartpole episode) and using only the last batch. As you mentioned, because you want to sample between each episode, a `EpisodeSampleRatioController` might be needed. It should not be too hard to implement (just a InsertSampleRatioController that counts inserts only when terminal = true). 

If you don't want to implement such a controller in a separate PR I understand. In that case you case also replace this selected piece of code with `batch = sample(trajectory)` to bypass the controller. Note that by doing this you lose some features such as the threshold of the controller and the possibility to adjust the number of episodes between samplings.

Reply:
Resolved in 033dcdf. I decided to use the second method